src/core/tsi/private_key_offload_util.cc

```diff
@@ -0,0 +1,155 @@
+//
+//
+// Copyright 2025 gRPC authors.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+//
+//
+
+#include "src/core/tsi/private_key_offload_util.h"
+
+#include <openssl/ssl.h>
+#include <sys/types.h>
+
+#include <cstddef>
+#include <cstdint>
+#include <string>
+#include <utility>
+
+#include "src/core/tsi/transport_security_interface.h"
+#include "absl/functional/bind_front.h"
+#include "absl/status/statusor.h"
+#include "absl/strings/string_view.h"
+
+namespace grpc_core {
+
+absl::StatusOr<PrivateKeySigner::SignatureAlgorithm> ToSignatureAlgorithmClass(
+    uint16_t algorithm) {
+#if defined(OPENSSL_IS_BORINGSSL)
+  switch (algorithm) {
+    case SSL_SIGN_RSA_PKCS1_SHA256:
+      return PrivateKeySigner::SignatureAlgorithm::kRsaPkcs1Sha256;
+    case SSL_SIGN_RSA_PKCS1_SHA384:
+      return PrivateKeySigner::SignatureAlgorithm::kRsaPkcs1Sha384;
+    case SSL_SIGN_RSA_PKCS1_SHA512:
+      return PrivateKeySigner::SignatureAlgorithm::kRsaPkcs1Sha512;
+    case SSL_SIGN_ECDSA_SECP256R1_SHA256:
+      return PrivateKeySigner::SignatureAlgorithm::kEcdsaSecp256r1Sha256;
+    case SSL_SIGN_ECDSA_SECP384R1_SHA384:
+      return PrivateKeySigner::SignatureAlgorithm::kEcdsaSecp384r1Sha384;
+    case SSL_SIGN_ECDSA_SECP521R1_SHA512:
+      return PrivateKeySigner::SignatureAlgorithm::kEcdsaSecp521r1Sha512;
+    case SSL_SIGN_RSA_PSS_RSAE_SHA256:
+      return PrivateKeySigner::SignatureAlgorithm::kRsaPssRsaeSha256;
+    case SSL_SIGN_RSA_PSS_RSAE_SHA384:
+      return PrivateKeySigner::SignatureAlgorithm::kRsaPssRsaeSha384;
+    case SSL_SIGN_RSA_PSS_RSAE_SHA512:
+      return PrivateKeySigner::SignatureAlgorithm::kRsaPssRsaeSha512;
+  }
+#endif  // OPENSSL_IS_BORINGSSL
+  return absl::InvalidArgumentError("Unknown signature algorithm.");
+}
+
+#if defined(OPENSSL_IS_BORINGSSL)
+void TlsOffloadSignDoneCallback(TlsPrivateKeyOffloadContext* ctx,
+                                absl::StatusOr<std::string> signed_data) {
+  ctx->status = TlsPrivateKeyOffloadContext::kCompleted;
+  if (signed_data.ok()) {
+    const uint8_t* bytes_to_send = nullptr;
+    size_t bytes_to_send_size = 0;
+    ctx->signed_bytes = std::move(signed_data);
+
+    tsi_result result = tsi_handshaker_next(
```

Comment:
Since this caused some trouble maybe add a brief comment about the wrapping here with `tsi_handshaker_next` and `notify_cb`